BUG : Fix Excel header NaN duplication with merged MultiIndex columns in to_excel

justine202429 opened the pull request:
- [x] closes #62340  
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
Fix Excel header NaN

Alvaro-Kothe reviewed (commented):
It needs tests and an entry in whatsnew.

mathbruu committed:
add test and update whatsnews

Alvaro-Kothe commented on pandas/tests/io/excel/test_writers.py:
```diff
@@ -1582,6 +1583,29 @@ def assert_called_and_reset(cls):
+        with ExcelWriter("test.xlsx", engine="openpyxl") as writer:
+            df.to_excel(writer, sheet_name="Sheet1", merge_cells=merge_cells)
+
+        reader = ExcelFile("test.xlsx")
```
Don't read and write to CWD, either use a temporary file, or use an in-memory buffer.

Alvaro-Kothe commented on pandas/tests/io/excel/test_writers.py:
```diff
@@ -1582,6 +1583,29 @@ def assert_called_and_reset(cls):
         DummyClass.assert_called_and_reset()
 
+    @td.skip_if_no("openpyxl")
+    def test_to_excel_multiindex_nan_in_columns(self, merge_cells):
```
Test all excel engines.

mathbruu commented:
pre-commit.ci autofix

mathbruu committed:
update test

mathbruu committed:
Merge branch 'main' into bugfix-Issue62340

mathbruu committed:
format file

Alvaro-Kothe commented on pandas/tests/io/excel/test_writers.py:
```diff
@@ -1507,6 +1507,24 @@ def test_to_excel_raising_warning_when_cell_character_exceed_limit(self):
+        with ExcelFile(tmp_excel) as reader:
+            result = pd.read_excel(reader, index_col=0, header=[0, 1])
+
+        tm.assert_numpy_array_equal(result.to_numpy(), df.to_numpy())
```
This doesn't test the header.

Commit:
BUG: Preserve NaN in MultiIndex column headers during Excel export

GH#62340: Use original column values (with NaN) instead of NBSP-filled
values when writing MultiIndex headers to Excel.

- Modify _format_header_mi() to use columns.get_level_values() to get
  the original column values with NaN preserved
- Add test to verify MultiIndex structure and data integrity are
  preserved during Excel round-trip
- Note: read_excel() limitation means NaN in headers become empty cells
  in Excel and cannot be reconstructed on read, but data values are
  correctly preserved

Commit:
STYLE: Fix whitespace in ExcelFormatter and TestExcelWriter for consistency

Commit:
fix format

jbrockmendel commented:
closing as stale. can reopen when you're ready to continue and address comments
